### DISCUSSION TITLE
What is the concept of clip?

### DISCUSSION THREAD

**Discussion by goddade** [ImpulseAdventure/GUIslice#532] [Q&A]
Hi, I want to port GUISlice to Arduino_GFX.
What is the role of clip in gui?
Can I update only the area of the rClipRect every time I update the screen?

**Comment by Pconti31** [comment]
@goddade this is only for the full screen size to avoid lines wrapping around the screen.
Most GUIslice drivers don't even need to clip since the harware driver in many cases ignores wrapping and clip automatically. 
you just need to add this routine to your new driver unless you are simply making mods to an existing driver.
```
bool gslc_DrvSetClipRect(gslc_tsGui* pGui,gslc_tsRect* pRect)
{
  // NOTE: The clipping rect is currently saved in the
  // driver struct, but the drawing code may or may not currently
  // use it.
  gslc_tsDriver*  pDriver = (gslc_tsDriver*)(pGui->pvDriver);
  if (pRect == NULL) {
    // Default to entire display
    pDriver->rClipRect = {0,0,pGui->nDispW,pGui->nDispH};
  } else {
    pDriver->rClipRect = *pRect;
  }
  return true;
}
```
Some drivers clip the line drawing lines others like TFT_eSPIdriver uses a view port to clip the screen, again some do nothing at all.
Paul--

**Reply by goddade** [reply]
This means that clip is not partial update.
Can GUISlice do partial update?
For example, only update a changed button.

**Comment by Pconti31** [comment]
@goddade Well, basically yes, but its more complicated than just one button.   GUIslice's core routines will take of that though. So, your driver doesn't need to take it into account.

For a more detailed answer on GUIslice's refresh code. 

The algorithm appears to be a simple variation of Apple's old package 
called QuickDraw refresh written for Apple Lisa. It was called **Regions** 
and was actually patented. The patent ran out back in 2003. The full 
algorithm is quite complex but GUIslice's implementation is not. 

NOTE: 
In talking with Calvin he was apparently unaware of the history of 
research in this area and came up with his implemention independently. 

GUIslice basic algorithm.
As UI Widgets mark themselves dirty or needing to be redrawn a flag is 
set in their data structure. Now all widgets are attached to a 'Page' 
inside a collection list. 

When a programmer calls update() function within Arduino's Loop function 
GUIslice goes through the current Page's Widget Collection list and 
looks at each widget, Any that are marked as needing redraw have their 
size added to a newly created invalidation region. 

This invalidation region is thus built up and gets larger and larger as 
more UI widgets in the page's container are processed. 

Once completed this invalid region gets cleared or erased on screen and 
then the UI widgets that are contained or fit in that invalid region are 
redrawn. All other widgets are ignored except for having their redraw 
flags turned off. 

Now this is more complicated than I outlined because GUIslice does 
support layers of pages but this should give you the general idea. 

Now we all know of modern methods to handle screen refresh like having a 
back store and double buffering and/or a region manager like Java Swing 
but given Calvin's routines fit on a Arduino UNO I say its pretty 
impressive. 

Paul--

**Comment by ImpulseAdventure** [comment]
Thank you @Pconti31 — you have summarized it well! Yes, I had independently come up with the redraw mechanism with the intent of keeping it relatively simple :)

@goddade - a port to Arduino_GFX would be really great! I have not spent much time looking at the API yet but hopefully it isn’t too much of a delta from the Adafruit GFX baseline.

**Comment by goddade** [comment]
@Pconti31 Thank you for your explanation. I want to use GUISlice on my previous e-paper project. E-paper cannot be updated frequently, and using partial updates has better performance.
@ImpulseAdventure I'm not familiar with Adafruit GFX, but from the [GUIslice_drv_adagfx.cpp](https://github.com/ImpulseAdventure/GUIslice/blob/master/src/GUIslice_drv_adagfx.cpp) file it seems that most of the APIs of Adafruit GFX and Arduino GFX are compatible.

**Comment by goddade** [comment]
@ImpulseAdventure 
Hi, I created a Pull request.#535
I tested some examples on ILI9341.
P.S. Modifying library directly in libdeps directory is really bad.

**Comment by ImpulseAdventure** [comment]
Hi @goddade --

Just thought I'd let you know that I have been working through your PR #535 . Thanks very much for posting this!

As you can imagine, creating a new version of the main "driver" (`GUIslice_drv_ardgfx.cpp`) was a good way to enable us to strip out a lot of the unused configuration code. On the other hand, it does mean that we would increase the library maintenance in applying mods to 4 drivers instead of 3 from this point onward.

So, in this past week I started looking into the changes in **GUIslice_drv_ardgfx** vs **GUIslice_drv_adagfx** and determining whether it may be practical to reuse the existing **GUIslice_drv_adagfx** with a new `DRV_DISP_ARDGFX` definition. Just curious whether you may have already tried doing that?

Other than the includes and device/bus interface instantiation, it looks like the main changes are in:
- gslc_DrvInit() (minor)
- gslc_DrvDrawTxt()
- gslc_DrvDrawBmp24FromMem() (minor)

Before I integrate the PR, I may continue to explore whether an enhancement to GUIslice_drv_ardgfx (based on your updates) might be feasible.

thanks!

**Reply by goddade** [reply]
@ImpulseAdventure 
As it stands, I think it's feasible to put them together. Actually, the only API difference I see is draw24bitRGBBitmap.
gslc_DrvDrawTxt should work using the same implementation as the Adafruit GFX driver, I modified it just because it seems cleaner this way.
Arduino GFX has two parts, a data bus related to the soc, and a GFX driver related to the screen. Data bus replaces SPI.
Arduino GFX supports U8g2 fonts, but it shouldn't require additional modifications.

**Reply by goddade** [reply]
The Arduino GFX can implement clip using canvas, which takes up extra memory.
I am trying to use GUISlice on E paper.
I want to update only the part after an elem change.
